### ISSUE TITLE
BOTH THEMES: Add custom pages capability to both themes.

### ISSUE THREAD

**Issue by jboz62** [meetinghouse/cms#217]
Currently, there is no function to add custom pages, although, Andy has pointed out that it exists in the code and needs to be activated. 

Comments from Andy:

For John and Barry - 
some rough instructions below - note that I don't expect these to work without a little tweaking, this is mostly just off the top of my head.  It should get you started though.  The laravel 4.2 documentation is excellent.

copy the contents of the app/views/pages/edit.blade.php file to existing app/views/pages/create.blade.php file.  remove any unnecessary fields (delete button, etc) and whatever other text changes you want.  line 13 should look something like ` {{ Form::model('page', array('method' => 'POST', 'route' => array('pages.store'), 'files' => 'true', 'role' => 'form')) }}` - note the minor changes there.  

Then in your `app/controllers/PagesController.php` file you should already have an empty `create` and `store` method.  The `create` method just needs to look like:
```
 parent::show();
        return View::make('pages.create');
```

and the `store` method should look something like:
```
$input = Input::all();
$validator = Validator::make($input, array('slug' => 'regex:/^\/[A-Za-z0-9_]+$/'));           
            if($validator->passes()) {
                $page = Page::create($input);
                $banner = $this->bannerSet($page);
                return Redirect::to("pages.index")->withMessage("Created Page");
            } else {
                return Redirect::to('pages/' . $page->id . '/edit')->withErrors($validator)
                    ->withMessage("Error ");
```

then you'll just need to add a button at the bottom of the `app/views/pages/index.blade.php` file like `<a class="btn btn-success" href="/pages/create">create page</a>`

**Comment by jboz62** [comment]
@acavanagh Barry and I have been looking through this, I have done a little digging on my own, as well. On the PostsContrtoller.php there is a  `public function adminIndex()` that is missing in the pages controller file, does that need to be added? The "create post" button that appears under "Admin Posts" comes from the admin_index.blade.php file, not the index.blade.php file. My guess is it needs to be the same for pages, am I right in that?

So far I've added the code as you provided it, trying different variations and added the admin_index.blade.php into apps/views/pages and added this code into the PagesContoller.php:
```
public function adminIndex()
	{
	parent::show();
	$pages = $this->pages->all();
	$banner = $this->banner;
		if($this->settings->theme == true)
		{
			return $this->respond($pages, 'pages.index_dark',  compact('pages', 'banner'));

		}else
		{
			return $this->respond($pages, 'pages.index',  compact('pages', 'banner'));
		}
	return View::make('pages.admin_index', compact('pages', 'settings'));
    }
```
Can you tell me if that code is correct and if that's the right approach?

Thanks
John

**Comment by acavanagh** [comment]
You will need both an admin_index route in teh routes.php file and an
admin_index method on the pages controller, as well as an admin_index blade
template in the pages templates folder.  They should all be very similar to
the way it's done in posts and projects.  What you have above looks about
right for the method in the controller.

**Comment by jboz62** [comment]
@acavanagh Btw, that code is on line 39 of the PagesContoller.php

**Comment by jboz62** [comment]
Ok, thanks, I'll take a look through those.

**Comment by jboz62** [comment]
@acavanagh Where do I find the routes.php file? I've not seen that one.

**Comment by jboz62** [comment]
@acavanagh Never mind, right in front of my nose ;)

**Comment by acavanagh** [comment]
app/routes.php.  It controls all the routing.  You can see routes like:
```
Route::get('/admin/posts', array(
  'before' => 'auth',
  'as'     => 'admin_posts',
  'uses'   => 'PostsController@adminIndex'
));
```

You'll need the same thing for pages, so like this:
```
Route::get('/admin/pages', array(
  'before' => 'auth',
  'as'     => 'admin_pages',
  'uses'   => 'PagesController@adminIndex'
));
```

then when you go to `/admin/pages` it will call the `adminIndex` method on
the `PagesController`, which will query the database, get all the pages,
and return the admin_index blade template

**Comment by jboz62** [comment]
Added to app/routes.php line 38:
```
Route::get('/admin/pages', array(
  'before' => 'auth',
  'as'     => 'admin_pages',
  'uses'   => 'PagesController@adminIndex'
));
```

**Comment by jboz62** [comment]
@acavanagh Got it, thanks, didn't see you posted that, but figured it out. There must be something else missing, still not getting anything.

**Comment by acavanagh** [comment]
what do you get when you go to /admin/pages?

**Comment by jboz62** [comment]
@acavanagh Just the same as before, the listing of pages with edit buttons, no "create" button. Btw, we are playing with this on the test site, if you care to take a look: http://testdeploy.corbettresearchgroup.com/

**Comment by acavanagh** [comment]
you'll need to add an edit button to the pages. adminIndex template

**Comment by jboz62** [comment]
Like this one? `<a id="page-id-{{$page->id}}" href="/pages/{{$page->id}}/edit">edit</a>`?

**Comment by acavanagh** [comment]
more like `<a class="btn btn-success" href="/pages/create">create page</a>`

**Comment by jboz62** [comment]
There already is one on there.

**Comment by jboz62** [comment]
```
<div class="col-md-9 column content">
    <table class="table table-hover table-striped table-condensed">
        <thead>
        <tr>
            <th>Title</th>
            <th>Published</th>
            <th>Edit</th>
        </tr>
        </thead>
        <tbody>
            @foreach($pages as $page)
                <tr>
                    <td><a href="{{$page->slug}}">{{{$page->title}}}</a></td>
                    <td class="published">
                        @if($page->published)
                        Published
                        @else
                        Unpublished
                        @endif
                    </td>
                    <td>
                        <a id="page-id-{{$page->id}}"
                           href="/pages/{{$page->id}}/edit">
                            edit
                        </a>
                    </td>
                </tr>
            @endforeach
        </tbody>
    </table>
    <a class="btn btn-success" href="/pages/create">Create Page</a>
</div>
```

**Comment by acavanagh** [comment]
It's not there when I view the source in the browser.  You have a different
admin index template for the dark theme.

**Comment by acavanagh** [comment]
also, you need to add something to the `create` method on the
PagesCotroller - like:
```
parent::show();
    parent::show();
        return View::make('pages.create');
```

and then you need something in your pages.create template.  You have one,
but there's not much there.

**Comment by jboz62** [comment]
Yes, I added that code in the PagesController.php file, we're good there. The pages.create has the edit.create that you said to copy/paste over.

I'm not sure I understand your comment about the admin index. Btw, I'm not working on Github, I'm working via FTP directly on the site, could be why you're not seeing this. We haven't found a way to edit any way else as we don't have a virtual machine in place.

**Comment by jboz62** [comment]
This is the path for the admin.index.blade: /testdeploy.corbettresearchgroup.com/app/views/pages/admin_index.blade.php

**Comment by acavanagh** [comment]
not sure how you're going to get your code integrated with the rest of the
sites.  It kind of relies on git.

re: the template - there's two templates.  The one you linked to is only
used by the light theme.  Looks like you're using the dark theme, which is
controlled by app/views/projects/admin_index_dark.blade.php

**Comment by acavanagh** [comment]
actually, that's not true for pages, just posts and projects.

**Comment by jboz62** [comment]
Yeah, we know that with the git reliance, figuring that out, Barry's on that, I think.

**Comment by jboz62** [comment]
> actually, that's not true for pages, just posts and projects.

Hmm, so that should be the right one then, right?

**Comment by acavanagh** [comment]
Check out your adminIndex method:
```
public function adminIndex()
{
parent::show();
$pages = $this->pages->all();
$banner = $this->banner;
if($this->settings->theme == true)
{
return $this->respond($pages, 'pages.index_dark',  compact('pages',
'banner'));

}else
{
return $this->respond($pages, 'pages.index',  compact('pages', 'banner'));
}
return View::make('pages.admin_index', compact('pages', 'settings'));
    }
```

I don't think that last line doesn't get hit because of the if else
statement.  You're either returning `$this->respond($pages,
'pages.index_dark',  compact('pages', 'banner'));` or `return
$this->respond($pages, 'pages.index',  compact('pages', 'banner'));`.  You
have the dark theme on, so it's the former.  It's looking for a
pages.index_dark
template.

**Comment by acavanagh** [comment]
If you want the same template for either theme, then what you want it to
look like is just:
```
$pages = $this->pages->all();
$banner = $this->banner;

return View::make('pages.admin_index', compact('pages', 'settings'));
    }
```

**Comment by jboz62** [comment]
Ah, that's actually what was in there before, I got fancy and messed it up. I'll revert to what you posted. Thanks for that!

**Comment by acavanagh** [comment]
no problem.  You might want a different template for the dark theme, in
which case you'll need something like the if/else statement, but at the
moment that's what's messing you up.  As soon as the method hits a `return`
it exits, so it never got to your last line.

**Comment by jboz62** [comment]
Is this line correct? `$pages = $this->pages->all();`? I changed it from what is in the posts controller, which is `$posts = Post::all();`. Wasn't sure about if it should be `Post::all();`.
```
public function adminIndex()
	{
	$pages = $this->pages->all();
	$banner = $this->banner;

	return View::make('pages.admin_index', compact('pages', 'settings'));
    }
```

**Comment by acavanagh** [comment]
Should work.

`$pages = $this->pages->all();` is equivalent to `Pages::all()`;
For whatever reason, you can see the `Pages` model being added to the
class's `$this->pages` property in the constructor at
https://github.com/meetinghouse/cms/blob/blog/app/controllers/PagesController.php#L15-15
.

**Comment by jboz62** [comment]
Ah, yes, ok, see that, thanks. I think I've exhausted my capabilities on this, if you see something I've missed or messed up, please let me know. Barry is taking a look at it now too, maybe he'll see something that isn't right. Thanks, Andy!

**Comment by acavanagh** [comment]
This is your problem:

the link in the top admin menu to `admin pages` goes to `/pages` not
`/admin/pages`, which is what you set up your route as.  Note how the links
for posts and projects are up there.  If you actually go to
http://testdeploy.corbettresearchgroup.com/admin/pages you'll hit your
method.  Currently that throws an error, but you just need to add
`parent::show()` to your `adminIndex` method on the `PagesController` and
then you should be all set.

**Comment by barryaprogrammer** [comment]
yea - we'll release thru git
i'm reading laravel doc as i follow along - thx

**Comment by jboz62** [comment]
Ah, I see that now, good catch. Let me see if I can make those changes, thanks!

**Comment by jboz62** [comment]
Yes, it's showing up now!

**Comment by jboz62** [comment]
Our create.blade is messed up, however, but this is a good start, for sure.

**Comment by jboz62** [comment]
Andy, can you point me to where the admin menu links are located?

**Comment by jboz62** [comment]
Found it. nav.blade.php

**Comment by jboz62** [comment]
Andy, we're getting an error on the create.blade for pages that says, "Undefined variable: page (View: /home/forge/testdeploy.corbettresearchgroup.com/app/views/pages/create.blade.php)" It has the `$page` variable in 2 places at the top the first is `<h2>Edit Page: {{$page->title}}</h2>`. Any idea how define the variable?

**Comment by acavanagh** [comment]
if you log out `$page` in the controller, what do you get?

**Comment by barryaprogrammer** [comment]
not sure what you mean to .. log out $page in the controller 

i am guessing we are not assiging a variable for page which is needed 
we have the debugger turned on now

**Comment by acavanagh** [comment]
You need to inspect the value of $page in the adminIndex method in the
pages controller.  You should be able to do that with your debugger, or
with `dd($page)` or `print_r($page)` or `Log::debug(print_r($page, 1))`

**Comment by jboz62** [comment]
Ok, thanks, Andy, we'll do that, going to take a break for some dinner. Thanks for the help on this!

**Comment by acavanagh** [comment]
No problem. You're almost there.

**Comment by ShaneBrenizer** [comment]
Can you guys take me off this thread? My inbox is filling up. Thanks

**Comment by jboz62** [comment]
Sorry about all the emails, Shane, not sure how you got added. 

Do we have a way to remove Shane from this thread? Looks to me the only way is for him to click the "unsubscribe" button that's under labels up top. Is there another way?

**Comment by jboz62** [comment]
> You need to inspect the value of $page in the adminIndex method in the
pages controller.  You should be able to do that with your debugger, or
with `dd($page)` or `print_r($page)` or `Log::debug(print_r($page, 1))`

Andy, the variable is actually `$pages` in the adminIndex method in the pages controller, but it doesn't work either way. However, it does print out the page arrays when you do a print_r, so that seems to be what you'd expect, right? 

So, I deleted some of the code in the create blade file, and now I get the display to create a new page, so that's good. Not sure what I did was right, but it works (I removed all the ` && $pages->id != 1` code and commented out the slideshow and delete button). However, when I fill in the fields and click "update", I get this error:
```
Symfony \ Component \ Debug \ Exception \ FatalErrorException
Access to undeclared static property: Page::$rules
```
The is in the page controller, line 110: `$rules = Page::$rules;`. Any idea what that means and how I might fix it? I'm not finding any useful info on the web so far.

**Comment by jboz62** [comment]
Ok, I found this post: http://stackoverflow.com/questions/25887647/access-to-undeclared-static-property-userrules-in-laravel and added the code below to the page model file. That seems to have fixed the above `Page::$rules` error.
```
public static $rules = array(
        'email' => 'required|email|unique:users',
        'password' => 'required|min:8',
    );
```
Now on to the next error, looks like the last, maybe. Now we're onto the database, adding the page. I'm getting this:
```
Illuminate \ Database \ QueryException
SQLSTATE[23000]: Integrity constraint violation: 19 NOT NULL constraint failed: pages.published (SQL: insert into "pages" ("title", "seo", "body", "slug", "updated_at", "created_at") values (Test Page, This is a test., <p>This is a test.</p> , /test, 2016-02-05 06:53:57, 2016-02-05 06:53:57))
```
The query exception occurred here: /vendor/laravel/framework/src/Illuminate/Database/Connection.php line 555
```
throw new QueryException($query, $this->prepareBindings($bindings), $e);
```
Some discussion here on similar topic: https://laracasts.com/discuss/channels/general-discussion/sql-sqlstate23000-integrity-constraint-violation Several other links as well, will have to do more reading. If you have any idea about this, Andy, please chime in, thanks!

**Comment by acavanagh** [comment]
you're on the right track with the $rules, but it looks like you
copy-pasted some that don't make sense for pages.  The $rules array defines
the validation rules for the various fields on the Pages mode.  The ones
that you copied over say that the 'email' field on pages is required and
has to be a valid email and should be unique in the users table, and that
the 'password' field is required and has to be at least 8 characters.
These would be great rules for the Users model, but there is no email nor
password field on the Pages model, so they don't make sense here.  See what
set of rules the 'edit' method for pages uses and apply those instead.

The query exception is happening because you're trying to save a model
without one of the required fields.  I can't tell from the error which one
it is, but checkout the 'store' method.  The following fields are all
required:
```
                'title' => "",
                'body'  => "",
                'published' => ,
                'slug' => '',
                'menu_sort_order' => ,
                'menu_parent' => ,
                'menu_name' => '',
```

Not sure what you have in the 'store' method atm, but if you dump the input
I'm willing to bet that you'll see that at least one of the above fields is
empty.  Possibly one of the menu fields.  You'll either have to add that
field to the create template and make it required, or provide a default in
the store method.

**Comment by jboz62** [comment]
> you're on the right track with the $rules, but it looks like you
copy-pasted some that don't make sense for pages.

Hah! That's not surprising, figured it wasn't right, but wanted to get past the error for now and didn't fully understand what that was, appreciate the clarification.

Great insight on the required fields, I'll look into that, sounds like we're almost there. I will take a look at the 'store' method and see what's in there. Thanks for your help, Andy!

**Comment by KathJohnson** [comment]
I see it is possible to create and view pages on the testdeploy site. Are there still errors preventing this from being rolled out? I am very excited to see this capability opening up so quickly; it will be really useful.

**Comment by jboz62** [comment]
It's almost there, Katherine. We're hung up on a menu error that is giving us a little trouble. Should be usable very soon. 


John Bozaan
Bozaan Design
413-437-0097

**Comment by KathJohnson** [comment]
Hey, folks. I know Barry is out for the moment, but I am checking in on this because it is urgent at this point. 

We will need this to be available in the light theme shortly as well.

**Comment by jboz62** [comment]
Hey Katherine,

I thought Barry already committed those changes to Github, might just need to be pushed out to the sites. I'll ask him the status and get back to you, giving him a ride to the airport in a bit.

John
--
John Bozaan
Bozaan Design
413-437-0097
www.bozaandesign.com

**Comment by meetinghouse** [comment]
Nope, I see it. I didn't even check the S+H2 site first, since I hadn't
heard from you! Thank you for your work (and Barry's). I'll start using it,
and if I don't encounter any problems, I'll close the issue. Many, many
thanks again, because this enables a lot more customizations.

I'll put the "next" label on the next-priority issue, which will be #167.

Sincerely,
Katherine Johnson, Research Director
Corbett Research Group
http://www.restorationtrades.com
"Web Marketing that Works"

Publishers of:

The Custom Building & Restoration Trades Directory
http://www.restorationtradesdirectory.com

The Building Arts Notebook, "Journal of the Guild of Building Artisans"
http://buildingartisansguild.com/

Email:  katherine@restorationtrades.com
Phone: 413-586-4748

**Comment by jboz62** [comment]
Ok, great! It's not perfect yet, but functional. We didn't add an image upload yet, but that shouldn't be a problem, if it's needed. 

--
John Bozaan
Bozaan Design
413-437-0097

**Comment by jboz62** [comment]
One other thing, the menus are still static. So for now, if you need a new menu item, just post it on git. 

--
John Bozaan
Bozaan Design
413-437-0097

**Comment by KathJohnson** [comment]
Oh, don't worry, I'll put in an issue for the menu management soon :)

But for now, I do have a problem with this issue too. I see a "Create" button on the testdeploy site, but it didn't make it into the codebase. I pushed code to this site: http://lightthemetest.corbettresearchgroup.com to test it, and while the Admin Pages is there in both light theme and dark theme, there is no create button, so I'm hamstrung. Can this be corrected urgently, particularly for the shconstruction2 site?

**Comment by jboz62** [comment]
The changes were committed to the "admin_pages_feature" branch, according to Barry. I think they needed to be pushed to "blog" and then deployed. Is that what you did? I just did a quick compare on the page controller file for "testdeploy" and "lightthemetest", there is missing code in "lightthemetest". That's probably true for the rest of the code as well. I'd review the changes in the "admin_pages_feature" branch and see if that's what needs to be pushed.

**Comment by acavanagh** [comment]
that branch needs to be merged into the blog branch and then deployed.  I
just made a pull request for that.  Changes can be reviewed here:
https://github.com/meetinghouse/cms/pull/220/files, and when you're ready
to merge you can click on the 'conversations' tab and press the merge
button.

**Comment by jboz62** [comment]
Thanks Andy!

**Comment by jboz62** [comment]
So I guess when we commit changes to a branch, this is the final step to take, create a pull request. I'll pass that along to Barry. Does the branch go away once it's merged?

I will be upgrading my OS so I can run git, haven't been able to install it so far. Once that is done, I'll commit changes for S+H that were overwritten.

John

**Comment by acavanagh** [comment]
You have the option to delete the branch after it's merged

**Comment by meetinghouse** [comment]
Just FYI, I didn't push code to the shconstruction2 site. I don't know the
exact status of code there, but just want to confirm that I've left it
untouched, so you may not need to worry about anything being overwritten.

Sincerely,
Katherine Johnson, Research Director
Corbett Research Group
http://www.restorationtrades.com
"Web Marketing that Works"

Publishers of:

The Custom Building & Restoration Trades Directory
http://www.restorationtradesdirectory.com

The Building Arts Notebook, "Journal of the Guild of Building Artisans"
http://buildingartisansguild.com/

Email:  katherine@restorationtrades.com
Phone: 413-586-4748

**Comment by jboz62** [comment]
Thanks, Andy.

Katherine, what was overwritten happened a while back, not recently. I just haven't bothered replacing it until I can do it through the right channel (git). Barry was going to do it, but he was using an editor that couldn't compare files well and wasn't able to figure it out in time before he left for Florida.

**Comment by jboz62** [comment]
Some of the formatting is missing on S+H and the mobile menu I added is also missing, I can ftp them up for now and then commit the changes when I get set up, if that works. Those changes should be in the base code, as well, for the dark theme, so I'll create a branch to add the changes and then we can do a merge. Should be set up by the weekend, I hope, if all goes well.

**Comment by KathJohnson** [comment]
That is great news. This will really be helpful, and I'll be implementing it right away for the S+H client, and actually will almost certainly be using it soon for a couple of light theme clients as well, one new and one established. I am working on a menu management issue now, as a next step from this item. Thanks for all your work on this, all.

**Comment by jboz62** [comment]
Katherine, I was looking at my code changes for the dark theme (on S+H) and realized, now that I understand how this system works better, that I need to add some code to separate dark and light theme changes. Go ahead and do the merge so I can utilize those changes and then I'll create a branch and add my dark theme changes.

**Comment by KathJohnson** [comment]
John, I don't know how to do a merge. Do you mean push the code from Git to the site? 

I see that Forge is down right now, so I can't push code at the moment. Gah!

**Comment by KathJohnson** [comment]
Thanks for taking my call John. I was able to merge branches, push code to the test site lightthemetest, and confirm the "Create Page" button in both light and dark themes. I've switched lightthemetest back to the light theme again, and pushed code to the shconstruction2 site. I'll be in that site creating and editing pages now, and if anything comes up, I'll note it.

Thanks for all your work, folks. I'm closing this issue now.
